src/ctap2/commands/mod.rs

```diff
@@ -103,136 +102,137 @@ pub(crate) enum PinUvAuthResult {
     SuccessGetPinToken,
+    /// Successfully established PinUvAuthToken via UV (CTAP2.1)
+    SuccessGetPinUvAuthTokenUsingUvWithPermissions,
+    /// Successfully established PinUvAuthToken via Pin (CTAP2.1)
+    SuccessGetPinUvAuthTokenUsingPinWithPermissions,
 }
 
 pub(crate) trait PinUvAuthCommand: RequestCtap2 {
     fn pin(&self) -> &Option<Pin>;
     fn set_pin(&mut self, pin: Option<Pin>);
     fn set_pin_auth(&mut self, pin_auth: Option<PinUvAuthParam>);
     fn client_data_hash(&self) -> ClientDataHash;
-    fn unset_uv_option(&mut self);
+    fn set_uv_option(&mut self, uv: Option<bool>);
+    fn get_uv_option(&mut self) -> Option<bool>;
+    fn get_rp_id(&self) -> Option<&String>;
 
     /// If Ok() is returned, the request can be sent to the device
     /// If Err() is returned, either cancel the operation or ask the user for a PIN
     fn determine_pin_uv_auth<D: FidoDevice>(
         &mut self,
         dev: &mut D,
         skip_uv: bool,
+        permission: PinUvAuthTokenPermission,
     ) -> Result<PinUvAuthResult, AuthenticatorError> {
+        // In case we error out in the middle
+        self.set_pin_auth(None);
+
         // CTAP1/U2F-only devices do not support PinUvAuth, so we skip it
         if !dev.supports_ctap2() {
             self.set_pin_auth(None);
             return Ok(PinUvAuthResult::DeviceIsCtap1);
         }
 
-        self.determine_pin_uv_auth_ctap_2_0(dev, skip_uv)
-    }
-
-    /// CTAP 2.0-only version:
-    /// "Getting pinUvAuthToken using getPinToken (superseded)"
-    fn determine_pin_uv_auth_ctap_2_0<D: FidoDevice>(
-        &mut self,
-        dev: &mut D,
-        skip_uv: bool,
-    ) -> Result<PinUvAuthResult, AuthenticatorError> {
         let info = dev
             .get_authenticator_info()
-            .ok_or(AuthenticatorError::Platform)?;
-
-        if self.pin().is_none() {
-            // Only use UV, if the device supports it and we don't skip it
-            // which happens as a fallback, if UV-usage failed too many times
-            let supports_uv = !skip_uv && info.options.user_verification == Some(true);
-            let supports_pin = info.options.client_pin.is_some();
-            let pin_configured = info.options.client_pin == Some(true);
+            .ok_or(AuthenticatorError::HIDError(HIDError::DeviceNotInitialized))?;
+
+        // Only use UV, if the device supports it and we don't skip it
+        // which happens as a fallback, if UV-usage failed too many times
+        // Note: In theory, we could also repeatedly query GetInfo here and check
+        //       if uv is set to Some(true), as tokens should set it to Some(false)
+        //       if UV is blocked (too many failed attempts). But the CTAP2.0-spec is
+        //       vague and I don't trust all tokens to implement it that way. So we
+        //       keep track of it ourselves, using `skip_uv`.
+        let supports_uv = !skip_uv && info.options.user_verification == Some(true);
```

Comment:
It's confusing to have `supports_uv` depends on `skip_uv`. This is one of the things I had changed in #239.

Reply:
Whoops, sorry. I missed that in the merge.